### PR TITLE
Fix race condition in trigger disable

### LINKED ISSUE #2587: Error EAI_AGAIN when initializing FCM
Hi, I am new to the firebase. I build a service using Node.js in order to do push notification activity. There is no problem when I test it in development. However, when it is in the production I found an error with initialization when I am trying to send a message. 

### [REQUIRED] Environment info

**firebase-tools:** FCM, npm firebase-admin ver.812.1

**Platform:** Linux Ubuntu

### [REQUIRED] Error

```
FirebaseAppError: Credential implementation provided to initializeApp() via the "credential" property failed to fetch a valid Google OAuth2 access token with the following error: "Error fetching access token: Error while making request: getaddrinfo EAI_AGAIN accounts.google.com. Error code: EAI_AGAIN".
    at FirebaseAppError.FirebaseError [as constructor] (/usr/src/node_modules/firebase-admin/lib/utils/error.js:42:28)
    at FirebaseAppError.PrefixedFirebaseError [as constructor] (/usr/src/node_modules/firebase-admin/lib/utils/error.js:88:28)
    at new FirebaseAppError (/usr/src/node_modules/firebase-admin/lib/utils/error.js:123:28)
    at /usr/src/node_modules/firebase-admin/lib/firebase-app.js:121:23
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async FirebaseMessagingCloudService.sendPushNotification (/usr/src/services/FirebaseMessagingCloudService.js:106:38)
    at async MessageService.createMessage (/usr/src/services/MessageService.js:144:13)
    at async /usr/src/controllers/MessageController/MessageController.js:52:28
    at async MessageController.module.exports [as parseMultipartForm] (/usr/src/controllers/parseMultipartForm.js:13:5)
    at async MessageController.createMessage (/usr/src/controllers/MessageController/MessageController.js:51:9) {
  errorInfo: {
    code: 'app/invalid-credential',
    message: 'Credential implementation provided to initializeApp() via the "credential" property failed to fetch a valid Google OAuth2 access token with the following error: "Error fetching access token: Error while making request: getaddrinfo EAI_AGAIN accounts.google.com. Error code: EAI_AGAIN".'
  },
  codePrefix: 'app'
}

```
here how I declare the serviceAccount:
```
const serviceAccount = {
    type: "service_account",
    project_id,
    private_key_id: process.env.FCM_PRIVATE_KEY_ID,
    private_key: process.env.DEVELOPMENT_ENV
        ? process.env.FCM_PRIVATE_KEY
        : process.env.FCM_PRIVATE_KEY.replace(/\\n/g, "\n"),
    client_email: `firebase-adminsdk-hb55c@${project_id}.iam.gserviceaccount.com`,
    client_id: process.env.FCM_PRIVATE_CLIENT_ID,
    auth_uri: "https://accounts.google.com/o/oauth2/auth",
    token_uri: "https://oauth2.googleapis.com/token",
    auth_provider_x509_cert_url: "https://www.googleapis.com/oauth2/v1/certs",
    client_x509_cert_url: process.env.FCM_PRIVATE_CERT_URL,
};

...
async initializeFCM() {
        try {
            await firebaseAdmin.initializeApp({
                credential: firebaseAdmin.credential.cert(serviceAccount),
                databaseURL: `https://${project_id}.firebaseio.com`,
            });
        } catch (error) {
            console.error(error);
            return INITIALIZATION_ERROR("FCM initialization");
        }
    }
...
```

### LINKED ISSUE #2857: The new feature withFunctionTriggersDisabled not really disable the triggers
* Operating System version: Linux / Ubuntu 20.04
  * Browser version: Node 12
  * Firebase SDK version: 8.0.2, @firebase/rules-unit-testing: 1.1.1, firebase-admin: 9.4.0
  * Firebase Product: functions

### The problem
Firebase 8.0.1 comes with a testing feature to *disable functions triggers* in emulator, I tried to use that new feature to clear *Firestore* data without triggering some triggers to make my tests more reliable, but that not worked, the triggers are called even wrapping custom firestore queries to delete the data manually

#### Steps to reproduce:
1. Init a project using firebase init
2. Set an actual project
3. Create a bunch of docs on a single collection (I used 'affiliates')
4. Make sure to use the real project name in initializeTestApp to allow triggers to run
5. After wrapping the clearFirestore code make the code wait some seconds the allow triggers to  run

#### Repository with test
https://github.com/saculbr/disableTriggersIssue

#### Relevant Code:
```javascript
test('disable triggers test', async () => {
    await firebase.withFunctionTriggersDisabled(async () => {
        return await firebase.clearFirestoreData({ projectId: APP_ID });
    });

     // sleep three seconds to allow triggers to run
     await sleep(3000);

     const doc = await db.collection('bar').doc('bar').get();
     expect(doc.data()).toBeUndefined()
}, 10000); // increase test timeout
```

```javascript
export const onDeleteAffiliate = functions
    .firestore.document(`affiliates/{affiliateId}`)
    .onDelete(async (docSnap, context) => {
        await sleep(100);
        await db.collection('bar').doc('bar').set({bar: 'bar'});
    });
```

### PR DESCRIPTION
**Pull request by samtstern**
### Description

Fixes #2857

We give all background triggers a "generation number" which allows us to disable triggers inside the functions emulator and then when incoming calls try to run those triggers, even after a large delay, we can ignore it.

### Scenarios Tested

functions/index.js
```js
const functions = require('firebase-functions');

exports.httpFunction = functions.https.onRequest((request, response) => {
  console.log("httpFunction");
  response.send("Hello from Firebase!");
});

exports.rtdbFunction = functions.database.ref('/foo/{bar}').onWrite((change, ctx) => {
  console.log("rtdbFunction: ", ctx.params.bar);
  return true;
});

exports.firestoreFunctions = functions.firestore.document('/foo/{bar}').onWrite((change, ctx) => {
  console.log("firestoreFunction: ", ctx.params.bar);
  return true;
});
```

functions/test.js
```js
const firebase = require("@firebase/rules-unit-testing");

async function addDoc(fs, id) {
  await fs.collection('foo').doc(id).set({
    date: new Date().getTime()
  });
  console.log("Added: ", id);
}

async function main() {
  const app = firebase.initializeAdminApp({
    projectId: 'fir-dumpster'
  });
  const fs = app.firestore();

  await addDoc(fs, 'should-trigger-1');
  await firebase.withFunctionTriggersDisabled(async () => {
    await addDoc(fs, 'should-not-trigger');
  });
  await addDoc(fs, 'should-trigger-2');
}

main();
```

Logs:
```
$ firebase emulators:exec "node test.js"                      
i  emulators: Starting emulators: functions, firestore, database
⚠  hub: emulator hub unable to start on port 4400, starting on 4401 instead.
⚠  functions: The following emulators are not running, calls to these services from the Functions emulator will affect production: auth, hosting, pubsub
⚠  Your requested "node" version "12" doesn't match your global version "10"
⚠  firestore: Did not find a Cloud Firestore rules file specified in a firebase.json config file.
⚠  firestore: The emulator will default to allowing all reads and writes. Learn more about this option: https://firebase.google.com/docs/emulator-suite/install_and_configure#security_rules_configuration.
i  firestore: Firestore Emulator logging to firestore-debug.log
⚠  database: Did not find a Realtime Database rules file specified in a firebase.json config file. The emulator will default to allowing all reads and writes. Learn more about this option: https://firebase.google.com/docs/emulator-suite/install_and_configure#security_rules_configuration.
i  database: Database Emulator logging to database-debug.log
i  functions: Watching "/private/var/folders/xl/6lkrzp7j07581mw8_4dlt3b000643s/T/tmp.EtFy933F/functions" for Cloud Functions...
✔  functions[httpFunction]: http function initialized (http://localhost:5001/fir-dumpster/us-central1/httpFunction).
✔  functions[rtdbFunction]: database function initialized.
✔  functions[firestoreFunctions]: firestore function initialized.
i  Running script: node test.js
Added:  should-trigger-1
i  emulators: Disabling Cloud Functions triggers, non-HTTP functions will not execute.
i  functions[rtdbFunction]: function temporarily disabled.
i  functions[firestoreFunctions]: function temporarily disabled.
Added:  should-not-trigger
i  emulators: Enabling Cloud Functions triggers, non-HTTP functions will execute.
i  functions: Beginning execution of "firestoreFunctions"
✔  functions[rtdbFunction]: database function initialized.
✔  functions[firestoreFunctions]: firestore function initialized.
Added:  should-trigger-2
>  firestoreFunction:  should-trigger-1
i  functions: Finished "firestoreFunctions" in ~1s
i  functions: Beginning execution of "firestoreFunctions"
>  firestoreFunction:  should-trigger-2
i  functions: Finished "firestoreFunctions" in ~1s
✔  Script exited successfully (code 0)
i  emulators: Shutting down emulators.
i  functions: Stopping Functions Emulator
i  database: Stopping Database Emulator
i  firestore: Stopping Firestore Emulator
i  hub: Stopping emulator hub
```

### Sample Commands

N/A

### CODE REVIEW

File: src/emulator/functionsEmulator.ts [hunk 1]
```diff
@@ -346,119 +352,13 @@ export class FunctionsEmulator implements EmulatorInstance {
       persistent: true,
     });
 
-    // TODO(abehaskins): Gracefully handle removal of deleted function definitions
```
Comment (samtstern) [inline src/emulator/functionsEmulator.ts:349]: This code was just moved out to a method

**Comment by samtstern** [comment]
Well clearly I broke something but I can't figure out what...

File: src/emulator/functionsEmulator.ts [hunk 2]
```diff
@@ -479,6 +381,110 @@ export class FunctionsEmulator implements EmulatorInstance {
...
+      worker.runtime.events,
+      "SYSTEM",
+      "triggers-parsed"
+    );
+    const triggerDefinitions = triggerParseEvent.data
+      .triggerDefinitions as EmulatedTriggerDefinition[];
+
+    const triggerResults: TriggerDescription[] = [];
+
+    const toSetup = triggerDefinitions.filter((definition) => !this.triggers[definition.name]);
```
Comment (samtstern) [inline src/emulator/functionsEmulator.ts:416]: The only change in this method is the removal of the 'knownTriggerIds' map

**Comment by samtstern** [comment]
@yuchenshi don't look at this one yet, I realized it's flawed.

**Comment by samtstern** [comment]
@yuchenshi this one is ready to go now.

**Review by yuchenshi — APPROVED** [review]
LGTM! One question: What happens if the functions code is hot-reloaded while trigger is disabled? Would it cause a race condition somehow and re-enable the functions early?

**Comment by samtstern** [comment]
@yuchenshi hot reload would not re-enable existing functions because the generation number won't be increased.

**Pull request merged**